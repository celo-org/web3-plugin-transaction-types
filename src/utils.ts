export enum TxTypeToPrefix {
  "ethereum-legacy" = "",
  cip42 = "0x7c",
  cip64 = "0x7b",
  cip66 = "0x7a",
  eip1559 = "0x02",
}

const celo = {
  chainId: 42220,
  rpcUrl: "https://forno.celo.org",
  explorerUrl: "https://celoscan.io",
} as const;
const alfajores = {
  chainId: 44787,
  rpcUrl: "https://alfajores-forno.celo-testnet.org",
  explorerUrl: "https://alfajores.celoscan.io",
} as const;

export const CeloChains = {
  celo,
  mainnet: celo,
  alfajores,
  testnet: alfajores,
} as const;
