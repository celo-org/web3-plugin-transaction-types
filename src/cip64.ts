import { keccak256 } from "ethereum-cryptography/keccak";
import { validateNoLeadingZeroes } from "web3-validator";
// eslint-disable-next-line import/no-extraneous-dependencies
import { RLP } from "@ethereumjs/rlp";
import {
  AccessList,
  AccessListUint8Array,
  BaseTransaction,
  Common,
  FeeMarketEIP1559TxData,
  JsonTx,
  TxOptions,
  TxValuesArray,
  bigIntToHex,
  bigIntToUnpaddedUint8Array,
  ecrecover,
  toUint8Array,
  txUtils,
  uint8ArrayToBigInt,
  unpadUint8Array,
} from "web3-eth-accounts";
import {
  bytesToHex,
  hexToBytes,
  uint8ArrayConcat,
  uint8ArrayEquals,
} from "web3-utils";
import { Address } from "web3";
import { TxTypeToPrefix } from "./utils";

const {
  getAccessListData,
  getAccessListJSON,
  getDataFeeEIP2930,
  verifyAccessList,
} = txUtils;

const MAX_INTEGER = BigInt(
  "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
);

const TRANSACTION_TYPE_UINT8ARRAY = hexToBytes(
  TxTypeToPrefix.cip64.padStart(2, "0")
);

type CIP64Data = FeeMarketEIP1559TxData & {
  feeCurrency: Address;
};
type CIP64ValuesArray = [
  Uint8Array,
  Uint8Array,
  Uint8Array,
  Uint8Array,
  Uint8Array,
  Uint8Array,
  Uint8Array,
  Uint8Array,
  AccessListUint8Array,
  Uint8Array,
  Uint8Array?,
  Uint8Array?,
  Uint8Array?
];
type CIP64JsonTx = JsonTx & {
  feeCurrency: string;
};

/**
 * Create a new EIP-2718 Transaction Type that will act as eip1559 transaction but including the feeCurrency field for paying in a different currency.
 *
 * - TransactionType: 0x7b
 * - CIP: [CIP64](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md)
 */
export class CIP64Transaction extends BaseTransaction<CIP64Transaction> {
  public readonly chainId: bigint;
  public readonly accessList: AccessListUint8Array;
  public readonly AccessListJSON: AccessList;
  public readonly maxPriorityFeePerGas: bigint;
  public readonly maxFeePerGas: bigint;
  public readonly feeCurrency: Address;

  public readonly common: Common;

  public static fromTxData(txData: CIP64Data, opts: TxOptions = {}) {
    return new CIP64Transaction(txData, opts);
  }

  public static fromSerializedTx(serialized: Uint8Array, opts: TxOptions = {}) {
    if (
      !uint8ArrayEquals(serialized.subarray(0, 1), TRANSACTION_TYPE_UINT8ARRAY)
    ) {
      throw new Error(
        `Invalid serialized tx input: not an CIP64 transaction (wrong tx type, expected: ${
          TxTypeToPrefix.cip64
        }, received: ${bytesToHex(serialized.subarray(0, 1))}`
      );
    }
    const values = RLP.decode(serialized.subarray(1));

    if (!Array.isArray(values)) {
      throw new Error("Invalid serialized tx input: must be array");
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    return CIP64Transaction.fromValuesArray(values as any, opts);
  }

  public static fromValuesArray(
    values: CIP64ValuesArray,
    opts: TxOptions = {}
  ) {
    if (values.length !== 10 && values.length !== 13) {
      throw new Error(
        "Invalid CIP64 transaction. Only expecting 10 values (for unsigned tx) or 13 values (for signed tx)."
      );
    }

    const [
      chainId,
      nonce,
      maxPriorityFeePerGas,
      maxFeePerGas,
      gasLimit,
      to,
      value,
      data,
      accessList,
      feeCurrency,
      v,
      r,
      s,
    ] = values;

    this._validateNotArray({ chainId, v });
    validateNoLeadingZeroes({
      nonce,
      maxPriorityFeePerGas,
      maxFeePerGas,
      gasLimit,
      value,
      v,
      r,
      s,
    });

    return new CIP64Transaction(
      {
        chainId: uint8ArrayToBigInt(chainId),
        nonce,
        maxPriorityFeePerGas,
        maxFeePerGas,
        gasLimit,
        to,
        value,
        data,
        accessList: accessList ?? [],
        feeCurrency: Buffer.from(feeCurrency).toString("hex"),
        v: v !== undefined ? uint8ArrayToBigInt(v) : undefined,
        r,
        s,
      },
      opts
    );
  }
  public constructor(txData: CIP64Data, opts: TxOptions = {}) {
    const {
      chainId,
      accessList,
      maxFeePerGas,
      maxPriorityFeePerGas,
      feeCurrency,
    } = txData;

    super({ ...txData, type: TxTypeToPrefix.cip64 }, opts);

    this.common = this._getCommon(opts.common, chainId);
    this.chainId = this.common.chainId();

    this.feeCurrency = feeCurrency;

    // Populate the access list fields
    const accessListData = getAccessListData(accessList ?? []);
    this.accessList = accessListData.accessList;
    this.AccessListJSON = accessListData.AccessListJSON;
    // Verify the access list format.
    verifyAccessList(this.accessList);

    this.maxFeePerGas = uint8ArrayToBigInt(
      toUint8Array(maxFeePerGas === "" ? "0x" : maxFeePerGas)
    );
    this.maxPriorityFeePerGas = uint8ArrayToBigInt(
      toUint8Array(maxPriorityFeePerGas === "" ? "0x" : maxPriorityFeePerGas)
    );

    this._validateCannotExceedMaxInteger({
      maxFeePerGas: this.maxFeePerGas,
      maxPriorityFeePerGas: this.maxPriorityFeePerGas,
    });

    BaseTransaction._validateNotArray(txData);

    if (this.gasLimit * this.maxFeePerGas > MAX_INTEGER) {
      const msg = this._errorMsg(
        "gasLimit * maxFeePerGas cannot exceed MAX_INTEGER (2^256-1)"
      );
      throw new Error(msg);
    }

    if (this.maxFeePerGas < this.maxPriorityFeePerGas) {
      const msg = this._errorMsg(
        "maxFeePerGas cannot be less than maxPriorityFeePerGas (The total must be the larger of the two)"
      );
      throw new Error(msg);
    }

    this._validateYParity();
    this._validateHighS();

    const freeze = opts?.freeze ?? true;
    if (freeze) {
      Object.freeze(this);
    }
  }

  public getDataFee(): bigint {
    if (
      this.cache.dataFee &&
      this.cache.dataFee.hardfork === this.common.hardfork()
    ) {
      return this.cache.dataFee.value;
    }

    let cost = super.getDataFee();
    cost += BigInt(getDataFeeEIP2930(this.accessList, this.common));

    if (Object.isFrozen(this)) {
      this.cache.dataFee = {
        value: cost,
        hardfork: this.common.hardfork(),
      };
    }

    return cost;
  }

  public getUpfrontCost(baseFee = BigInt(0)): bigint {
    const prio = this.maxPriorityFeePerGas;
    const maxBase = this.maxFeePerGas - baseFee;
    const inclusionFeePerGas = prio < maxBase ? prio : maxBase;
    const gasPrice = inclusionFeePerGas + baseFee;
    return this.gasLimit * gasPrice + this.value;
  }

  public raw(): TxValuesArray {
    return [
      bigIntToUnpaddedUint8Array(this.chainId),
      bigIntToUnpaddedUint8Array(this.nonce),
      bigIntToUnpaddedUint8Array(this.maxPriorityFeePerGas),
      bigIntToUnpaddedUint8Array(this.maxFeePerGas),
      bigIntToUnpaddedUint8Array(this.gasLimit),
      this.to !== undefined ? this.to.buf : Uint8Array.from([]),
      bigIntToUnpaddedUint8Array(this.value),
      this.data,
      this.accessList,
      hexToBytes(this.feeCurrency),
      this.v !== undefined
        ? bigIntToUnpaddedUint8Array(this.v)
        : Uint8Array.from([]),
      this.r !== undefined
        ? bigIntToUnpaddedUint8Array(this.r)
        : Uint8Array.from([]),
      this.s !== undefined
        ? bigIntToUnpaddedUint8Array(this.s)
        : Uint8Array.from([]),
    ] as TxValuesArray;
  }

  public serialize(): Uint8Array {
    const base = this.raw();
    return uint8ArrayConcat(TRANSACTION_TYPE_UINT8ARRAY, RLP.encode(base));
  }

  public getMessageToSign(hashMessage = true): Uint8Array {
    const base = this.raw().slice(0, 10);
    const message = uint8ArrayConcat(
      TRANSACTION_TYPE_UINT8ARRAY,
      RLP.encode(base)
    );
    if (hashMessage) {
      return keccak256(message);
    }
    return message;
  }

  public hash(): Uint8Array {
    if (!this.isSigned()) {
      const msg = this._errorMsg(
        "Cannot call hash method if transaction is not signed"
      );
      throw new Error(msg);
    }

    if (Object.isFrozen(this)) {
      if (!this.cache.hash) {
        this.cache.hash = keccak256(this.serialize());
      }
      return this.cache.hash;
    }
    return keccak256(this.serialize());
  }

  public getMessageToVerifySignature(): Uint8Array {
    return this.getMessageToSign();
  }

  public getSenderPublicKey(): Uint8Array {
    if (!this.isSigned()) {
      const msg = this._errorMsg(
        "Cannot call this method if transaction is not signed"
      );
      throw new Error(msg);
    }

    const msgHash = this.getMessageToVerifySignature();
    const { v, r, s } = this;

    this._validateHighS();

    try {
      return ecrecover(
        msgHash,
        v! + BigInt(27), // Recover the 27 which was stripped from ecsign
        bigIntToUnpaddedUint8Array(r!),
        bigIntToUnpaddedUint8Array(s!)
      );
    } catch (e: any) {
      const msg = this._errorMsg("Invalid Signature");
      throw new Error(msg);
    }
  }

  public _processSignature(v: bigint, r: Uint8Array, s: Uint8Array) {
    const opts = { ...this.txOptions, common: this.common };

    return CIP64Transaction.fromTxData(
      {
        chainId: this.chainId,
        nonce: this.nonce,
        maxPriorityFeePerGas: this.maxPriorityFeePerGas,
        maxFeePerGas: this.maxFeePerGas,
        gasLimit: this.gasLimit,
        to: this.to,
        value: this.value,
        data: this.data,
        accessList: this.accessList,
        feeCurrency: this.feeCurrency,
        v: v - BigInt(27), // This looks extremely hacky: /util actually adds 27 to the value, the recovery bit is either 0 or 1.
        r: uint8ArrayToBigInt(r),
        s: uint8ArrayToBigInt(s),
      },
      opts
    );
  }

  public toJSON(): CIP64JsonTx {
    const accessListJSON = getAccessListJSON(this.accessList);

    return {
      chainId: bigIntToHex(this.chainId),
      nonce: bigIntToHex(this.nonce),
      maxPriorityFeePerGas: bigIntToHex(this.maxPriorityFeePerGas),
      maxFeePerGas: bigIntToHex(this.maxFeePerGas),
      gasLimit: bigIntToHex(this.gasLimit),
      to: this.to !== undefined ? this.to.toString() : undefined,
      value: bigIntToHex(this.value),
      data: bytesToHex(this.data),
      accessList: accessListJSON,
      feeCurrency: this.feeCurrency,
      v: this.v !== undefined ? bigIntToHex(this.v) : undefined,
      r: this.r !== undefined ? bigIntToHex(this.r) : undefined,
      s: this.s !== undefined ? bigIntToHex(this.s) : undefined,
    };
  }

  public errorStr() {
    let errorStr = this._getSharedErrorPostfix();
    errorStr += ` maxFeePerGas=${this.maxFeePerGas} maxPriorityFeePerGas=${this.maxPriorityFeePerGas}`;
    return errorStr;
  }

  protected _errorMsg(msg: string) {
    return `${msg} (${this.errorStr()})`;
  }
}
