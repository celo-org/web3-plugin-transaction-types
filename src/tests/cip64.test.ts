import { describe, it, expect } from "bun:test";
import { RLP } from "@ethereumjs/rlp";
import { hexToBytes, toWei } from "web3-utils";

import { CIP64Transaction } from "../cip64";
import {
  Common,
  privateKeyToAddress,
  privateKeyToPublicKey,
} from "web3-eth-accounts";
import { CeloChains } from "../utils";

const common = Common.custom({
  // using the same as in
  // https://github.com/celo-org/developer-tooling/blob/master/packages/sdk/wallets/wallet-local/src/local-wallet.test.ts
  chainId: 44378,
});

const TWO_POW256 = BigInt(
  "0x10000000000000000000000000000000000000000000000000000000000000000"
);

const PRIVATE_KEY1 =
  "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";
const PUBLIC_KEY1 = privateKeyToPublicKey(PRIVATE_KEY1, false);
const ACCOUNT_ADDRESS1 = privateKeyToAddress(PRIVATE_KEY1);
const PRIVATE_KEY2 =
  "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890fdeccc";
const ACCOUNT_ADDRESS2 = privateKeyToAddress(PRIVATE_KEY2);

const baseTxData = {
  from: ACCOUNT_ADDRESS1,
  to: ACCOUNT_ADDRESS2,
  chainId: `0x${(44378).toString(16)}`,
  value: BigInt(toWei(1, "ether")),
  nonce: 0,
  gasLimit: BigInt("10"),
  gas: BigInt("10"),
  data: "0xabcdef",
  maxFeePerGas: BigInt("99"),
  maxPriorityFeePerGas: BigInt("99"),
  feeCurrency: "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826",
};

describe("[CIP64Transaction]", () => {
  it("sign()", () => {
    const pkey = hexToBytes(PRIVATE_KEY1);
    const txn = CIP64Transaction.fromTxData(baseTxData, { common });
    const signed = txn.sign(pkey);
    const rlpSerialized = RLP.encode(Uint8Array.from(signed.serialize()));
    expect(rlpSerialized).toMatchSnapshot();
  });

  it("hash()", () => {
    const pkey = hexToBytes(PRIVATE_KEY1);
    let txn = CIP64Transaction.fromTxData(baseTxData, { common });
    let signed = txn.sign(pkey);
    const expectedHash = hexToBytes(
      // Copied from https://github.com/celo-org/developer-tooling/blob/master/packages/sdk/wallets/wallet-local/src/local-wallet.test.ts#L320
      "0x645afc1d19fe805c0c0956e70d5415487bf073741d7b297ccb7e7040c6ce5df6"
    );
    expect(signed.hash()).toEqual(expectedHash);
  });

  it("fromSerializedTx()", () => {
    let txn = CIP64Transaction.fromSerializedTx(
      hexToBytes(
        // Copied from https://github.com/celo-org/developer-tooling/blob/master/packages/sdk/wallets/wallet-local/src/local-wallet.test.ts#L315
        "0x7bf88282ad5a8063630a94588e4b68193001e4d10928660ab4165b813717c0880de0b6b3a764000083abcdefc094cd2a3d9f938e13cd947ec05abc7fe734df8dd82680a091b5504a59e529e7efa42dbb97fbc3311a91d035c873a94ab0789441fc989f84a02e8254d6b3101b63417e5d496833bc84f4832d4a8bf8a2b83e291d8f38c0f62d"
      ),
      { common }
    );
    expect(txn.toJSON()).toMatchSnapshot();
  });

  it("toJSON()", () => {
    const pkey = hexToBytes(PRIVATE_KEY1);
    const txn = CIP64Transaction.fromTxData(
      { ...baseTxData, to: ACCOUNT_ADDRESS2 },
      { common }
    );
    const signed = txn.sign(pkey);

    const json = signed.toJSON();
    const expectedJSON = {
      chainId: "0xad5a",
      nonce: "0x0",
      maxPriorityFeePerGas: "0x63",
      maxFeePerGas: "0x63",
      gasLimit: "0xa",
      to: "0x588e4b68193001e4d10928660ab4165b813717c0",
      value: "0xde0b6b3a7640000",
      data: "0xabcdef",
      accessList: [],
      feeCurrency: "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826",
      v: "0x0",
      r: "0x91b5504a59e529e7efa42dbb97fbc3311a91d035c873a94ab0789441fc989f84",
      s: "0x2e8254d6b3101b63417e5d496833bc84f4832d4a8bf8a2b83e291d8f38c0f62d",
    };
    expect(json).toEqual(expectedJSON);
  });

  it("Fee validation", () => {
    expect(() => {
      CIP64Transaction.fromTxData(
        {
          ...baseTxData,
          maxFeePerGas: TWO_POW256 - BigInt(1),
          maxPriorityFeePerGas: 100,
          gasLimit: 1,
          value: 6,
        },
        { common }
      );
    }).not.toThrow();
    expect(() => {
      CIP64Transaction.fromTxData(
        {
          ...baseTxData,
          maxFeePerGas: TWO_POW256 - BigInt(1),
          maxPriorityFeePerGas: 100,
          gasLimit: 100,
          value: 6,
        },
        { common }
      );
    }).toThrow();
    expect(() => {
      CIP64Transaction.fromTxData(
        {
          ...baseTxData,
          maxFeePerGas: 1,
          maxPriorityFeePerGas: 2,
          gasLimit: 100,
          value: 6,
        },
        { common }
      );
    }).toThrow();
  });
});
