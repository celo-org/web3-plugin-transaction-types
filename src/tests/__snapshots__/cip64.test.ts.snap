// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`[CIP64Transaction] sign() 1`] = `
Uint8Array [
  184,
  133,
  123,
  248,
  130,
  130,
  173,
  90,
  128,
  99,
  99,
  10,
  148,
  88,
  142,
  75,
  104,
  25,
  48,
  1,
  228,
  209,
  9,
  40,
  102,
  10,
  180,
  22,
  91,
  129,
  55,
  23,
  192,
  136,
  13,
  224,
  182,
  179,
  167,
  100,
  0,
  0,
  131,
  171,
  205,
  239,
  192,
  148,
  205,
  42,
  61,
  159,
  147,
  142,
  19,
  205,
  148,
  126,
  192,
  90,
  188,
  127,
  231,
  52,
  223,
  141,
  216,
  38,
  128,
  160,
  145,
  181,
  80,
  74,
  89,
  229,
  41,
  231,
  239,
  164,
  45,
  187,
  151,
  251,
  195,
  49,
  26,
  145,
  208,
  53,
  200,
  115,
  169,
  74,
  176,
  120,
  148,
  65,
  252,
  152,
  159,
  132,
  160,
  46,
  130,
  84,
  214,
  179,
  16,
  27,
  99,
  65,
  126,
  93,
  73,
  104,
  51,
  188,
  132,
  244,
  131,
  45,
  74,
  139,
  248,
  162,
  184,
  62,
  41,
  29,
  143,
  56,
  192,
  246,
  45,
]
`;

exports[`[CIP64Transaction] fromSerializedTx() 1`] = `
{
  "accessList": [],
  "chainId": "0xad5a",
  "data": "0xabcdef",
  "feeCurrency": "cd2a3d9f938e13cd947ec05abc7fe734df8dd826",
  "gasLimit": "0xa",
  "maxFeePerGas": "0x63",
  "maxPriorityFeePerGas": "0x63",
  "nonce": "0x0",
  "r": "0x91b5504a59e529e7efa42dbb97fbc3311a91d035c873a94ab0789441fc989f84",
  "s": "0x2e8254d6b3101b63417e5d496833bc84f4832d4a8bf8a2b83e291d8f38c0f62d",
  "to": "0x588e4b68193001e4d10928660ab4165b813717c0",
  "v": "0x0",
  "value": "0xde0b6b3a7640000",
}
`;
